### ISSUE TITLE
Subcribed client doesn't ping the server while receiving messages, QoS=0 (no ack).

### ISSUE THREAD

**Issue by AlmostRuio** [vert-x3/vertx-mqtt#190]
Vertx 4.0.0

### Context

The MqttClient seems to stop pinging while receiving messages with QoS=0 (no ack). The client is disconnected after keep alive is expired.

### Reproducer

https://github.com/AlmostRuio/vertx-mqtt-timeout

#### Steps  to reproduce

1. Start a local broker (e.g. [eclipse-mosquitto Docker Official Images](https://hub.docker.com/_/eclipse-mosquitto/))
1. Build the reproducer project and run the Main
1. Publish message every second (with e.g. mosquitto_sub) to the subscibed topic (see command below in Logs)
1. The client stop pinging and will be disconnected after 10 second from the broker (KeepAliveInterval set in MqttClient)

### Logs

Here we see what happend in the logs:
1. Vertx MqttClient -> stop pinging while receiving messages QoS=0 (no ack)
1. Broker -> show connection and timeout
1. Publisher -> publish messages every seconds
1. Working Client -> continue pinging while receiving messages. It will not be disconnected (see command below in Logs)

#### Vertx MqttClient 4.0.0
```
2020-12-15T12:58:04.293+0100 DEBUG [vert.x-eventloop-thread-1] MqttClientImpl - Trying to connect with localhost:1883
2020-12-15T12:58:04.427+0100 INFO [vert.x-eventloop-thread-1] MqttClientImpl - Connection with localhost:1883 established successfully
2020-12-15T12:58:04.440+0100 DEBUG [vert.x-eventloop-thread-1] MqttClientImpl - Sending packet MqttConnectMessage[fixedHeader=MqttFixedHeader[messageType=CONNECT, isDup=false, qosLevel=AT_MOST_ONCE, isRetain=false, remainingLength=0], variableHeader=MqttConnectVariableHeader[name=MQTT, version=4, hasUserName=false, hasPassword=false, isWillRetain=false, isWillFlag=false, isCleanSession=true, keepAliveTimeSeconds=10], payload=MqttConnectPayload[clientIdentifier=e0b7bfc1-adf5-467a-9bf7-a27617b8d03e, willTopic=null, willMessage=null, userName=null, password=null]]
2020-12-15T12:58:04.461+0100 DEBUG [vert.x-eventloop-thread-1] MqttClientImpl - Incoming packet MqttConnAckMessage[fixedHeader=MqttFixedHeader[messageType=CONNACK, isDup=false, qosLevel=AT_MOST_ONCE, isRetain=false, remainingLength=2], variableHeader=MqttConnAckVariableHeader[connectReturnCode=CONNECTION_ACCEPTED, sessionPresent=false], payload=]
2020-12-15T12:58:04.465+0100 DEBUG [vert.x-eventloop-thread-1] MqttClientImpl - Sending packet MqttSubscribeMessage[fixedHeader=MqttFixedHeader[messageType=SUBSCRIBE, isDup=false, qosLevel=AT_LEAST_ONCE, isRetain=false, remainingLength=0], variableHeader=MqttMessageIdVariableHeader[messageId=1], payload=MqttSubscribePayload[MqttTopicSubscription[topicFilter=test/topic, qualityOfService=AT_MOST_ONCE]]]
2020-12-15T12:58:04.466+0100 INFO [vert.x-eventloop-thread-1] ClientMqttVerticle - connected to localhost:1883, subscribed topic "test/topic"
2020-12-15T12:58:04.468+0100 DEBUG [vert.x-eventloop-thread-1] MqttClientImpl - Incoming packet MqttSubAckMessage[fixedHeader=MqttFixedHeader[messageType=SUBACK, isDup=false, qosLevel=AT_MOST_ONCE, isRetain=false, remainingLength=3], variableHeader=MqttMessageIdVariableHeader[messageId=1], payload=MqttSubAckPayload[grantedQoSLevels=[0]]]
2020-12-15T12:58:14.475+0100 DEBUG [vert.x-eventloop-thread-1] MqttClientImpl - Sending packet MqttMessage[fixedHeader=MqttFixedHeader[messageType=PINGREQ, isDup=false, qosLevel=AT_MOST_ONCE, isRetain=false, remainingLength=0], variableHeader=, payload=]
2020-12-15T12:58:14.477+0100 DEBUG [vert.x-eventloop-thread-1] MqttClientImpl - Incoming packet MqttMessage[fixedHeader=MqttFixedHeader[messageType=PINGRESP, isDup=false, qosLevel=AT_MOST_ONCE, isRetain=false, remainingLength=0], variableHeader=, payload=]
2020-12-15T12:58:20.006+0100 DEBUG [vert.x-eventloop-thread-1] MqttClientImpl - Incoming packet MqttPublishMessage[fixedHeader=MqttFixedHeader[messageType=PUBLISH, isDup=false, qosLevel=AT_MOST_ONCE, isRetain=false, remainingLength=28], variableHeader=MqttPublishVariableHeader[topicName=test/topic, packetId=-1], payload=PooledSlicedByteBuf(ridx: 0, widx: 16, cap: 16/16, unwrapped: PooledUnsafeDirectByteBuf(ridx: 30, widx: 30, cap: 512))]
2020-12-15T12:58:21.001+0100 DEBUG [vert.x-eventloop-thread-1] MqttClientImpl - Incoming packet MqttPublishMessage[fixedHeader=MqttFixedHeader[messageType=PUBLISH, isDup=false, qosLevel=AT_MOST_ONCE, isRetain=false, remainingLength=28], variableHeader=MqttPublishVariableHeader[topicName=test/topic, packetId=-1], payload=PooledSlicedByteBuf(ridx: 0, widx: 16, cap: 16/16, unwrapped: PooledUnsafeDirectByteBuf(ridx: 30, widx: 30, cap: 496))]
2020-12-15T12:58:22.002+0100 DEBUG [vert.x-eventloop-thread-1] MqttClientImpl - Incoming packet MqttPublishMessage[fixedHeader=MqttFixedHeader[messageType=PUBLISH, isDup=false, qosLevel=AT_MOST_ONCE, isRetain=false, remainingLength=28], variableHeader=MqttPublishVariableHeader[topicName=test/topic, packetId=-1], payload=PooledSlicedByteBuf(ridx: 0, widx: 16, cap: 16/16, unwrapped: PooledUnsafeDirectByteBuf(ridx: 30, widx: 30, cap: 496))]
2020-12-15T12:58:23.003+0100 DEBUG [vert.x-eventloop-thread-1] MqttClientImpl - Incoming packet MqttPublishMessage[fixedHeader=MqttFixedHeader[messageType=PUBLISH, isDup=false, qosLevel=AT_MOST_ONCE, isRetain=false, remainingLength=28], variableHeader=MqttPublishVariableHeader[topicName=test/topic, packetId=-1], payload=PooledSlicedByteBuf(ridx: 0, widx: 16, cap: 16/16, unwrapped: PooledUnsafeDirectByteBuf(ridx: 30, widx: 30, cap: 480))]
2020-12-15T12:58:24.004+0100 DEBUG [vert.x-eventloop-thread-1] MqttClientImpl - Incoming packet MqttPublishMessage[fixedHeader=MqttFixedHeader[messageType=PUBLISH, isDup=false, qosLevel=AT_MOST_ONCE, isRetain=false, remainingLength=28], variableHeader=MqttPublishVariableHeader[topicName=test/topic, packetId=-1], payload=PooledSlicedByteBuf(ridx: 0, widx: 16, cap: 16/16, unwrapped: PooledUnsafeDirectByteBuf(ridx: 30, widx: 30, cap: 480))]
2020-12-15T12:58:25.005+0100 DEBUG [vert.x-eventloop-thread-1] MqttClientImpl - Incoming packet MqttPublishMessage[fixedHeader=MqttFixedHeader[messageType=PUBLISH, isDup=false, qosLevel=AT_MOST_ONCE, isRetain=false, remainingLength=28], variableHeader=MqttPublishVariableHeader[topicName=test/topic, packetId=-1], payload=PooledSlicedByteBuf(ridx: 0, widx: 16, cap: 16/16, unwrapped: PooledUnsafeDirectByteBuf(ridx: 30, widx: 30, cap: 464))]
2020-12-15T12:58:26.006+0100 DEBUG [vert.x-eventloop-thread-1] MqttClientImpl - Incoming packet MqttPublishMessage[fixedHeader=MqttFixedHeader[messageType=PUBLISH, isDup=false, qosLevel=AT_MOST_ONCE, isRetain=false, remainingLength=28], variableHeader=MqttPublishVariableHeader[topicName=test/topic, packetId=-1], payload=PooledSlicedByteBuf(ridx: 0, widx: 16, cap: 16/16, unwrapped: PooledUnsafeDirectByteBuf(ridx: 30, widx: 30, cap: 464))]
2020-12-15T12:58:27.006+0100 DEBUG [vert.x-eventloop-thread-1] MqttClientImpl - Incoming packet MqttPublishMessage[fixedHeader=MqttFixedHeader[messageType=PUBLISH, isDup=false, qosLevel=AT_MOST_ONCE, isRetain=false, remainingLength=28], variableHeader=MqttPublishVariableHeader[topicName=test/topic, packetId=-1], payload=PooledSlicedByteBuf(ridx: 0, widx: 16, cap: 16/16, unwrapped: PooledUnsafeDirectByteBuf(ridx: 30, widx: 30, cap: 448))]
2020-12-15T12:58:28.008+0100 DEBUG [vert.x-eventloop-thread-1] MqttClientImpl - Incoming packet MqttPublishMessage[fixedHeader=MqttFixedHeader[messageType=PUBLISH, isDup=false, qosLevel=AT_MOST_ONCE, isRetain=false, remainingLength=28], variableHeader=MqttPublishVariableHeader[topicName=test/topic, packetId=-1], payload=PooledSlicedByteBuf(ridx: 0, widx: 16, cap: 16/16, unwrapped: PooledUnsafeDirectByteBuf(ridx: 30, widx: 30, cap: 448))]
2020-12-15T12:58:29.009+0100 DEBUG [vert.x-eventloop-thread-1] MqttClientImpl - Incoming packet MqttPublishMessage[fixedHeader=MqttFixedHeader[messageType=PUBLISH, isDup=false, qosLevel=AT_MOST_ONCE, isRetain=false, remainingLength=28], variableHeader=MqttPublishVariableHeader[topicName=test/topic, packetId=-1], payload=PooledSlicedByteBuf(ridx: 0, widx: 16, cap: 16/16, unwrapped: PooledUnsafeDirectByteBuf(ridx: 30, widx: 30, cap: 432))]
2020-12-15T12:58:30.011+0100 DEBUG [vert.x-eventloop-thread-1] MqttClientImpl - Incoming packet MqttPublishMessage[fixedHeader=MqttFixedHeader[messageType=PUBLISH, isDup=false, qosLevel=AT_MOST_ONCE, isRetain=false, remainingLength=28], variableHeader=MqttPublishVariableHeader[topicName=test/topic, packetId=-1], payload=PooledSlicedByteBuf(ridx: 0, widx: 16, cap: 16/16, unwrapped: PooledUnsafeDirectByteBuf(ridx: 30, widx: 30, cap: 432))]
2020-12-15T12:58:30.114+0100 WARN [vert.x-eventloop-thread-1] ClientMqttVerticle - connection to broker closed, reconnect
2020-12-15T12:58:30.125+0100 DEBUG [vert.x-eventloop-thread-1] MqttClientImpl - Trying to connect with localhost:1883
2020-12-15T12:58:30.129+0100 INFO [vert.x-eventloop-thread-1] MqttClientImpl - Connection with localhost:1883 established successfully
```

#### Broker (eclipse-mosquitto docker image 1.6.12)

```
1608031068: New connection from 172.20.0.1 on port 1883.
1608031068: New client connected from 172.20.0.1 as e0462c7a-b916-42bb-8a39-b426c22526ea (p2, c1, k10).
1608031124: Client e0462c7a-b916-42bb-8a39-b426c22526ea has exceeded timeout, disconnecting.
```

#### Publisher (mosquitto_pub 2.0.2)

```
/ $ mosquitto_pub -m "{\"status\":\"off\"}" -t "test/topic" --repeat 100 --repeat-delay 1 -p 1883
```

#### Working Client  (mosquitto_sub 2.0.2)

```
/ $ mosquitto_sub -t "test/topic" -p 1883 -v -d -k 10

Client (null) sending CONNECT
Client (null) received CONNACK (0)
Client (null) sending SUBSCRIBE (Mid: 1, Topic: test/topic, QoS: 0, Options: 0x00)
Client (null) received SUBACK
Subscribed (mid: 1): 0
Client (null) sending PINGREQ
Client (null) received PINGRESP
Client (null) received PUBLISH (d0, q0, r0, m0, 'test/topic', ... (16 bytes))
test/topic {"status":"off"}
Client (null) received PUBLISH (d0, q0, r0, m0, 'test/topic', ... (16 bytes))
test/topic {"status":"off"}
Client (null) received PUBLISH (d0, q0, r0, m0, 'test/topic', ... (16 bytes))
test/topic {"status":"off"}
Client (null) received PUBLISH (d0, q0, r0, m0, 'test/topic', ... (16 bytes))
test/topic {"status":"off"}
Client (null) received PUBLISH (d0, q0, r0, m0, 'test/topic', ... (16 bytes))
test/topic {"status":"off"}
Client (null) sending PINGREQ
Client (null) received PUBLISH (d0, q0, r0, m0, 'test/topic', ... (16 bytes))
test/topic {"status":"off"}
Client (null) received PINGRESP
Client (null) received PUBLISH (d0, q0, r0, m0, 'test/topic', ... (16 bytes))
test/topic {"status":"off"}
Client (null) received PUBLISH (d0, q0, r0, m0, 'test/topic', ... (16 bytes))
test/topic {"status":"off"}
Client (null) received PUBLISH (d0, q0, r0, m0, 'test/topic', ... (16 bytes))
test/topic {"status":"off"}
Client (null) received PUBLISH (d0, q0, r0, m0, 'test/topic', ... (16 bytes))
```

**Comment by vietj** [comment]
this has been fixed in master, can you give a try with latest 4.0.1-SNAPSHOT version ? https://github.com/vert-x3/vertx-mqtt/issues/189

**Comment by vietj** [comment]
tagging duplicate as suspected to be a duplicate

**Comment by AlmostRuio** [comment]
> this has been fixed in master, can you give a try with latest 4.0.1-SNAPSHOT version ? #189



> this has been fixed in master, can you give a try with latest 4.0.1-SNAPSHOT version ? #189

**Comment by AlmostRuio** [comment]
> this has been fixed in master, can you give a try with latest 4.0.1-SNAPSHOT version ? #189

I found the snapshot only here https://oss.sonatype.org/content/repositories/snapshots/io/vertx/vertx-mqtt/4.0.1-SNAPSHOT/
But the latest build is from 07.12.2020, so your fix isn't in it. (Tested)

Alternatively i can build my self the whole project but i don't know which project(s) to clone.
I didn't find any documentation.

Thanks anyway :)

**Comment by vietj** [comment]
yes please, we have an issue with CI that does not pass currently.

**Comment by AlmostRuio** [comment]
@vietj some how I built successfully the 4.0.1-SNAPSHOT.
I can confirm, the bug is **fixed** and is a duplicate of #189.

Thank you and have a nice one.
